frontend/src/metabase/lib/react-compat.ts

```diff
@@ -0,0 +1,47 @@
+// Support React 17 backwards compatibility for the Embedding SDK
+
+import React from "react";
+import ReactDOM from "react-dom";
+import { createRoot, type Root } from "react-dom/client";
+
+export function isReact17OrEarlier(version = React.version) {
+  const versionParts = version.split(".");
+
+  if (versionParts.length < 1) {
+    return false;
+  }
+
+  const majorVersion = parseInt(versionParts[0], 10);
+
+  if (isNaN(majorVersion)) {
+    return false;
+  }
+
+  return majorVersion <= 17;
+}
+
+export function renderRoot(
+  content: React.JSX.Element,
+  element: Element,
+): Root | undefined {
+  if (isReact17OrEarlier()) {
+    ReactDOM.render(content, element);
+    return;
+  }
+
+  const root = createRoot(element);
```

Comment:
Using `typeof createRoot === "function"` does not work, as that somehow also returns "true" even in React 17. Using `isReact17OrEarlier` seems to be the most reliable.

Reply:
should we use the same shim check? That the hook exists. Instead of this

Reply:
@ranquild This `renderRoot` is for a separate issue though, not related to the `useSyncExternalStore` hook - we already use `createRoot` in `TableInteractive`. Not sure what you exactly mean by checking if the hook exists.

Reply:
I mean, if you want to check if it's React 18, you can check for the hook

Reply:
@ranquild Oh right, `!("useSyncExternalStore" in React)` - good call!